go/vt/vtgate/engine/concatenate.go

```diff
@@ -101,22 +119,19 @@ func (c *Concatenate) TryExecute(vcursor VCursor, bindVars map[string]*querypb.B
 }
 
 func (c *Concatenate) getFields(res []*sqltypes.Result) ([]*querypb.Field, error) {
-	var resFields []*querypb.Field
+	if len(res) == 0 {
+		return nil, nil
+	}
+	fields := res[0].Fields
 	for _, r := range res {
-		fields := r.Fields
-		if fields == nil {
-			continue
-		}
-		if resFields == nil {
-			resFields = fields
-			continue
-		}
-		err := compareFields(fields, resFields)
-		if err != nil {
-			return nil, err
+		if r.Fields != nil {
+			err := c.compareFields(fields, r.Fields)
+			if err != nil {
+				return nil, err
+			}
```

Comment:
The previous version of the code did not assume that the first result would always have a field set! The new code however assigns fields to `res[0].Fields` which might be nil. The next res could have a non empty field and then we will end up comparing nil and non-nil fields which will throw an error.
Previously this would have worked. Is this change in behaviour intentional or a side effect of the refactor which needs fixing?

Reply:
Good point